components/zcash_note_encryption/src/lib.rs

```diff
@@ -1,8 +1,471 @@
+        &self,
+        cv: &D::ValueCommitment,
+        cm: &D::NoteCommitment,
+        rng: &mut R,
+    ) -> [u8; OUT_CIPHERTEXT_SIZE] {
+        let (ock, input) = if let Some(ovk) = &self.ovk {
+            let ock = D::get_ock(ovk, &cv, &cm, &self.epk);
+            let input = D::outgoing_plaintext_bytes(&self.note, &self.esk);
+
+            (ock, input)
+        } else {
+            // ovk = ⊥
+            let mut ock = OutgoingCipherKey([0; 32]);
+            let mut input = [0u8; OUT_PLAINTEXT_SIZE];
+
+            rng.fill_bytes(&mut ock.0);
+            rng.fill_bytes(&mut input);
+
+            (ock, OutPlaintextBytes(input))
+        };
+
+        let mut output = [0u8; OUT_CIPHERTEXT_SIZE];
+        assert_eq!(
+            ChachaPolyIetf::aead_cipher()
+                .seal_to(&mut output, &input.0, &[], ock.as_ref(), &[0u8; 12])
+                .unwrap(),
+            OUT_CIPHERTEXT_SIZE
+        );
+
+        output
+    }
+}
+
+/// Trial decryption of the full note plaintext by the recipient.
+///
+/// Attempts to decrypt and validate the given `enc_ciphertext` using the given `ivk`.
+/// If successful, the corresponding Sapling note and memo are returned, along with the
+/// `PaymentAddress` to which the note was sent.
+///
+/// Implements section 4.17.2 of the Zcash Protocol Specification.
+pub fn try_note_decryption<D: Domain>(
+    domain: &D,
+    //output: &ShieldedOutput<D>,
+    ivk: &D::IncomingViewingKey,
+    epk: &D::EphemeralPublicKey,
+    cmstar: &D::ExtractedCommitment,
+    enc_ciphertext: &[u8],
+) -> Option<(D::Note, D::Recipient, D::Memo)> {
+    assert_eq!(enc_ciphertext.len(), ENC_CIPHERTEXT_SIZE);
+
+    let shared_secret = D::ka_agree_dec(ivk, epk);
+    let key = D::kdf(shared_secret, epk);
+
+    let mut plaintext = [0; ENC_CIPHERTEXT_SIZE];
+    assert_eq!(
+        ChachaPolyIetf::aead_cipher()
+            .open_to(
+                &mut plaintext,
+                &enc_ciphertext,
+                &[],
+                key.as_ref(),
+                &[0u8; 12]
+            )
+            .ok()?,
+        NOTE_PLAINTEXT_SIZE
+    );
+
+    let (note, to) = parse_note_plaintext_without_memo_ivk(domain, ivk, epk, cmstar, &plaintext)?;
+    let memo = domain.extract_memo(&plaintext);
+
+    Some((note, to, memo))
+}
+
+fn parse_note_plaintext_without_memo_ivk<D: Domain>(
+    domain: &D,
+    ivk: &D::IncomingViewingKey,
+    epk: &D::EphemeralPublicKey,
+    cmstar: &D::ExtractedCommitment,
+    plaintext: &[u8],
+) -> Option<(D::Note, D::Recipient)> {
+    let (note, to) = domain.parse_note_plaintext_without_memo_ivk(ivk, &plaintext)?;
+
+    if let NoteValidity::Valid = check_note_validity::<D>(&note, epk, cmstar) {
+        Some((note, to))
+    } else {
+        None
+    }
+}
+
+fn check_note_validity<D: Domain>(
+    note: &D::Note,
+    epk: &D::EphemeralPublicKey,
+    cmstar: &D::ExtractedCommitment,
+) -> NoteValidity {
+    if &D::extract_note_commitment(&note) != cmstar {
```

Comment:
Technically, the spec says to reencode the extracted commitment (first rejecting ⊥), and compare that to the byte sequence `cm∗`. It's ok to compare it as an `ExtractedCommitment` provided that decoding to that type rejects non-canonical integer encodings.

Reply:
So should `extract_note_commitment` return a `NoteCommitment` instead, and then place a bound `: TryFrom<Self::NoteCommitment>` on `ExtractedCommitment`? I think that would be how to follow the spec more closely.

Reply:
In the current PR, `D::ExtractedCommitment` _is_ the encoded extracted commitment (for the Sapling impl it's `[u8; 32]`, literally computed as `note.cmu().to_bytes()`). ⊥ isn't rejected here because it doesn't occur for Sapling, but in the PR that adds Orchard support `D::extract_note_comitment` would be altered to return e.g. `Option<D::ExtractedCommitment>`.

Reply:
@nuttycom That seems reasonable; then it's `D::note_commitment(note: &Self::Note) -> Self::NoteCommitment` followed by a `try_into()` that Rust can constrain the output type of to match `cmstar`.

Reply:
Fixed in 0763d2b113